File: 2024/Day 24/Python/main.py
"""Advent of Code - 2024 - Day 24"""

import re

REG = dict[str, bool]
GATES = dict[str, tuple[str, str, str]]


def get_val(reg: REG, gates: GATES, name: str) -> int:
    """get value of gate"""
    if name in reg:
        return reg[name]
    (l, r, operand) = gates[name]
    val1 = get_val(reg, gates, l)
    val2 = get_val(reg, gates, r)
    if operand == "AND":
        return val1 & val2
    if operand == "OR":
        return val1 | val2
    return val1 ^ val2


def solve(reg: REG, gates: GATES, zgates: list[str]):
    """solve"""
    num = 0
    for operand in zgates:
        num = (num << 1) + get_val(reg, gates, operand)
    return num


def get_swap(gates: GATES, name: str) -> str:
    """gets the swap"""
    if name[0] == "z":
        return f"z{int(name[1:])-1:02d}"
    for out, (l, r, op) in gates.items():
        if name in (l, r) and op != "AND":
            return get_swap(gates, out)
    assert False


def get_badgates(reg: REG, gates: GATES, zgates: list[str]):
    """finds the bad gates"""
    rule1, rule2 = get_rules(gates, zgates)

    for rule in rule2:
        swap = get_swap(gates, rule)
        gates[rule], gates[swap] = gates[swap], gates[rule]

    num = solve(reg, gates, zgates)
    truenum = get_truenum(reg)
    trailing = ((num ^ truenum) & -(num ^ truenum)).bit_length() - 1
    badgates = rule1 + rule2
    for gate, (a, b, _) in gates.items():
        if set((a, b)) == set((f"x{trailing:02d}", f"y{trailing:02d}")):
            badgates.append(gate)
    return badgates


def get_truenum(reg: REG):
    """calculates the true result"""
    xnum: int = 0
    ynum: int = 0
    for name in sorted(reg, reverse=True):
        if name[0] == "x":
            xnum = (xnum << 1) + reg[name]
        if name[0] == "y":
            ynum = (ynum << 1) + reg[name]
    truenum = xnum + ynum
    return truenum


def get_rules(gates: GATES, zgates: list[str]):
    """finds the gates that violate rule 1 and 2"""
    rule1: list[str] = []
    rule2: list[str] = []
    for gate, (a, b, op) in gates.items():
        if gate in zgates[1:] and op != "XOR":
            rule1.append(gate)
        if gate not in zgates:
            if (
                a[0] != "x"
                and a[0] != "y"
                and b[0] != "x"
                and b[0] != "y"
                and op == "XOR"
            ):
                rule2.append(gate)
    return rule1, rule2


def main():
    """main function"""
    with open(0, encoding="utf-8") as f:
        p, q = f.read().split("\n\n")
    reg: REG = {line[:3]: line[5:] == "1" for line in p.splitlines()}
    gates: GATES = {}
    for line in q.splitlines():
        m = re.fullmatch("(...) (AND|OR|XOR) (...) -> (...)", line)
        if m:
            (a, op, b, c) = m.groups()
            gates[c] = (a, b, op)

    zgates: list[str] = sorted(filter(lambda gate: gate[0] == "z", gates), reverse=True)

    print(solve(reg, gates, zgates))
    print(",".join(sorted(get_badgates(reg, gates, zgates))))


main()

File: 2024/Day 24/C++/main.cpp
#include <string>
#include <iostream>
#include <unordered_map>
#include <tuple>
#include <regex>
#include <vector>

using REG = std::unordered_map<std::string, bool>;
using GATES = std::unordered_map<std::string, std::tuple<std::string, std::string, std::string>>;

bool get_val(const REG &reg, const GATES &gates, const std::string &name)
{
    if (reg.find(name) != reg.end())
    {
        return reg.at(name);
    }
    auto [l, r, operand] = gates.at(name);
    int l_val = get_val(reg, gates, l);
    int r_val = get_val(reg, gates, r);
    if (operand == "AND")
    {
        return l_val & r_val;
    }
    else if (operand == "or")
    {
        return l_val | r_val;
    }
    else
    {
        return l_val ^ r_val;
    }
}

long long solve(const REG &reg, const GATES &gates, const std::vector<std::string> &zgates)
{
    long long num = 0;
    for (auto &gate : zgates)
    {
        num = (num << 1) + get_val(reg, gates, gate);
    }
    return num;
}

std::string getswap(const GATES &gates, std::string name)
{
    if (name[0] == 'z')
    {
        std::string s = name.substr(1);
        int num = stoi(s) - 1;
        if (num < 10)
        {
            return "z0" + std::to_string(num);
        }
        return "z" + std::to_string(num);
    }
    for (auto &[out, v] : gates)
    {
        auto &[l, r, op] = v;
        if ((name == l || name == r) && op != "AND")
        {
            return getswap(gates, out);
        }
    }
    std::cout << "ERROR" << "\n";
    return "";
}

std::pair<std::vector<std::string>, std::vector<std::string>> get_rules(const GATES &gates, const std::vector<std::string> &zgates)
{
    std::vector<std::string> rule1;
    std::vector<std::string> rule2;
    for (auto &[gate, v] : gates)
    {
        auto &[l, r, op] = v;
        if (std::find(zgates.begin() + 1, zgates.end(), gate) != zgates.end() && op != "XOR")
        {
            rule1.push_back(gate);
        }
        if (std::find(zgates.begin(), zgates.end(), gate) == zgates.end())
        {
            if (l[0] != 'x' && l[0] != 'y' && r[0] != 'x' && r[0] != 'y' && op == "XOR")
            {
                rule2.push_back(gate);
            }
        }
    }
    return {rule1, rule2};
}

long long get_truenum(const REG &reg)
{
    std::vector<std::string> reg_names;
    for (auto &[name, _] : reg)
    {
        reg_names.push_back(name);
    }
    std::sort(reg_names.begin(), reg_names.end(), std::greater<std::string>());

    long long xnum = 0;
    long long ynum = 0;
    
    for (auto name : reg_names)
    {
        if (name[0] == 'x')
        {
            xnum = (xnum << 1) + reg.at(name);
        }
        else if (name[0] == 'y')
        {
            ynum = (ynum << 1) + reg.at(name);
        }
    }
    return xnum + ynum;
}

std::vector<std::string> get_badgates(const REG &reg, GATES &gates, const std::vector<std::string> &zgates)
{
    auto [rule1, rule2] = get_rules(gates, zgates);

    for (auto &rule : rule2)
    {
        std::string swap = getswap(gates, rule);
        std::swap(gates[rule], gates[swap]);
    }

    long long num = solve(reg, gates, zgates);
    long long truenum = get_truenum(reg);
    long long diff = ((num ^ truenum) & -(num ^ truenum));
    unsigned bits;
    for (bits = 0; diff != 0; bits++)
    {
        diff >>= 1;
    }
    bits--;
    std::vector<std::string> badgates;
    badgates.insert(badgates.end(), rule1.begin(), rule1.end());
    badgates.insert(badgates.end(), rule2.begin(), rule2.end());
    std::string xstr = "x" + std::to_string(bits);
    std::string ystr = "y" + std::to_string(bits);
    for (auto &[gate, v] : gates)
    {
        auto &[l, r, _] = v;
        if ((l == xstr && r == ystr) || (r == xstr && l == ystr))
        {
            badgates.push_back(gate);
        }
    }
    return badgates;
}

int main(int argc, char *argv[])
{
    std::string line;

    REG reg;
    while (std::getline(std::cin, line))
    {
        if (line == "")
        {
            break;
        }
        reg[line.substr(0, 3)] = line.substr(5, 1) == "1";
    }

    GATES gates;
    std::regex r("(...) (AND|OR|XOR) (...) -> (...)");
    while (std::getline(std::cin, line))
    {
        std::smatch m;
        if (std::regex_match(line, m, r))
        {
            gates[m[4]] = {m[1], m[3], m[2]};
        }
    }
    std::vector<std::string> zgates;
    for (auto &[k, _] : gates)
    {
        if (k[0] == 'z')
        {
            zgates.push_back(k);
        }
    }
    std::sort(zgates.begin(), zgates.end(), std::greater<std::string>());
    std::cout << solve(reg, gates, zgates) << "\n";

    std::vector<std::string> badgates = get_badgates(reg, gates, zgates);
    std::sort(badgates.begin(), badgates.end());
    for (auto &gate : badgates)
    {
        std::cout << gate;
        if (gate == badgates.back())
        {
            std::cout << "\n";
        }
        else
        {
            std::cout << ",";
        }
        
        
    }
    
    
}